pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,707 @@
+				&reserves,
+				T::Precision::get(),
+				amplification as Balance,
+				Balance::zero(),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			ensure!(
+				share_amount >= T::MinimumLiquidity::get(),
+				Error::<T>::InsufficientLiquidity
+			);
+
+			let pool_account = T::ShareAccountId::from_assets(&pool.assets, Some(POOL_IDENTIFIER));
+
+			T::Currency::transfer(assets.0, &who, &pool_account, initial_liquidity.0)?;
+			T::Currency::transfer(assets.1, &who, &pool_account, initial_liquidity.1)?;
+
+			T::Currency::deposit(pool_id.0, &who, share_amount)?;
+
+			Self::deposit_event(Event::PoolCreated {
+				id: pool_id,
+				assets,
+				initial_liquidity,
+				amplification,
+			});
+
+			Ok(())
+		}
+
+		/// Add liquidity to selected 2-asset pool.
+		///
+		/// LP must provide liquidity of both assets by specifying amount of asset a.
+		///
+		/// Amount of asset b is calculated so the ratio does not change:
+		///
+		/// new_reserve_b = (reserve_a + amount) * reserve_b / reserve_a
+		/// amount_b = new_reserve_b - reserve_b
+		///
+		/// LP must have sufficient amount of asset a/b - amount_a and amount_b.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: origin of the caller
+		/// - `pool_id`: Pool Id
+		/// - `asset`: Asset id
+		/// - `amount`: liquidity amount of `asset` to be added to the pool.
+		///
+		/// Emits `LiquidityAdded` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T::AssetId>,
+			asset: T::AssetId,
+			amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinimumTradingLimit::get(),
+				Error::<T>::InsufficientTradingAmount
+			);
+
+			// Ensure that who has enough balance of given asset
+			ensure!(
+				T::Currency::free_balance(asset, &who) >= amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+
+			ensure!(pool.contains_asset(asset), Error::<T>::AssetNotInPool);
+
+			// Load initial pool assets balances.
+			let pool_account = T::ShareAccountId::from_assets(&pool.assets, Some(POOL_IDENTIFIER));
+			let initial_reserves = AssetAmounts(
+				T::Currency::free_balance(pool.assets.0, &pool_account),
+				T::Currency::free_balance(pool.assets.1, &pool_account),
+			);
+
+			// Work out which asset's amount has to be calculated based on given provided asset by LP
+			// Calculate correct amount of second asset which LP has to provided too.
+			// Update initial reserves.
+			let (asset_b_id, asset_b_amount, new_reserves) = if asset == pool.assets.0 {
+				let asset_reserve = initial_reserves
+					.0
+					.checked_add(amount)
+					.ok_or(ArithmeticError::Overflow)?;
+
+				let asset_b_reserve = calculate_asset_b_reserve(initial_reserves.1, initial_reserves.0, asset_reserve)
```

Comment:
Given that assets and reserves are ordered, this scenario is handled correctly. line 427 handles a case where given assets is first asset, and line 444 handles scenario where provided assets is second asset.

Reply:
to be clear, I was talking about arguments in `calculate_asset_b_reserve`.

Reply:
i believe it is correct. 

the resulting new reserves are in correct order.

Reply:
Let's take for example a case where `asset == pool.assets.0`. In this case `pool.assets.0` is asset_a and `pool.assets.1` is asset_b. But we use `initial_reserves.1` in place of `asset_a_reserve` parameter.

Reply:
it is correct. However the naming of variables are unfortunate and it can be misleading. 

the `calculate_asset_b_reserve` actually  calculates new reserve for the first asset ( which is named 'a`). 

i will refactor this part.

Reply:
It was correct but misleading. This part has been refactored to be more readable.